File: DataStructures/Java/Stack.java
import java.util.ArrayList;

class Stack {
    private int top = 0;
    private final int maxSize;

    private ArrayList<Integer> stack;

    Stack(final int size) {
        this.maxSize = size;
        this.stack = new ArrayList<Integer>(this.maxSize);
    }

    public void push(final int value) {
        if (this.top == this.maxSize) {
            System.out.println("Stack Overflow");
            return;
        }

        this.stack.add(this.top, value);
        this.top++;
    }

    public int pop() {
        if (this.top == 0) {
            System.out.println("Stack Underflow");
            return -1;
        }

        this.top--;
        return this.stack.remove(this.top);
    }

    public int peek() {
        if (this.top == 0) {
            System.out.println("Stack Underflow");
            return -1;
        }

        return this.stack.get(this.top - 1);
    }

    public boolean isEmpty() {
        return this.top == 0;
    }

    public boolean isFull() {
        return this.top == this.maxSize;
    }

    public void print() {
        System.out.print("Stack: ");
        for (int i = 0; i < this.top; i++) {
            System.out.print(this.stack.get(i) + " ");
        }
        System.out.println();
    }

    public static void main(final String[] args) {
        Stack stack = new Stack(5);

        stack.push(1);
        stack.push(2);
        stack.push(3);
        stack.push(4);
        stack.print(); // Output: 1 2 3 4

        stack.push(5);
        stack.print(); // Output: 1 2 3 4 5

        stack.push(6); // Output: Stack Overflow
        stack.print(); // Output: 1 2 3 4 5

        System.out.println("Peek: " + stack.peek()); // Output: 5
        stack.print(); // Output: 1 2 3 4 5

        System.out.println("Pop: " + stack.pop()); // Output: 5
        stack.print(); // Output: 1 2 3 4
    }
}

File: DataStructures/CPP/stack.cpp
#include <iostream>
#include <vector>

using namespace std;

class Stack
{
public:
    int size;
    int top = -1;
    vector<int> stack;

    explicit Stack(int n)
    {
        this->size = n;
    }

    int push(int element)
    {
        if (this->top == this->size - 1)
        {
            cout << ("Stack Overflow\n");
            return -1;
        }

        this->top++;
        this->stack.push_back(element);

        return 0;
    }

    int pop()
    {
        if (this->top == -1)
        {
            cout << ("Stack Underflow\n");
            return -1;
        }

        this->top--;

        return 0;
    }

    int peek()
    {
        if (this->top == -1)
        {
            cout << ("Stack Underflow\n");
            return -1;
        }

        return this->stack[top];
    }

    void display()
    {
        if (this->top == -1)
        {
            cout << ("Stack Underflow\n");
            return;
        }

        cout << ("The stack is: ");
        for (int i = 0; i <= this->top; i++)
        {
            cout << (this->stack[i]) << " ";
        }
        cout << ("\n");
    }
};

int main()
{
    Stack s = Stack(5);
    s.push(3);
    s.push(4);
    s.push(5);
    s.pop();

    cout << s.peek() << endl;
    s.display();

    return 0;
}
